### PR TITLE
ipython: load js the IPython 3 way

### PR DESCRIPTION
**Pull request by cgestes**
no more js injection.

### CODE REVIEW

**Comment by rossant** [comment]
still WIP, doesn't work yet

**Comment by rossant** [comment]
Status update: making the JS code asynchronous doesn't work yet because the first custom message sent via comm arrives *before* a custom message handler is registered in the frontend. One work-around may be to implement a little handshake so that the frontend tells the backend when it's ready to process custom messages. cc @ctaf42

**Comment by rossant** [comment]
ipynb webgl doesn't work remotely for now because of the DPI that fails
http://stackoverflow.com/a/15916364/1595060

we should implement a handshake at the beginning where the widget sends to the backend a message with its cap + DPI

**Comment by cgestes** [comment]
VispyWidget is first class now.

The loading of the widget is fully asynchronous and follow ipython 3 rules

File: examples/ipynb/webgl_example_1.ipynb [hunk 1]
```diff
@@ -150,7 +150,10 @@
    "outputs": [],
    "source": [
     "c = Canvas(size=(300, 300))\n",
-    "c.show()"
+    "from vispy.app.backends.ipython import VispyWidget\n",
+    "w = VispyWidget()\n",
+    "w.set_canvas(c)\n",
+    "w"
```
Comment (rossant) [inline examples/ipynb/webgl_example_1.ipynb]: can we make a shortcut such that `c.show()` does this?

File: vispy/app/backends/_ipynb_webgl.py [hunk 2]
```diff
@@ -135,13 +130,20 @@ def __init__(self, *args, **kwargs):
...
 
-        self._create_widget(size=p.size)
+        #store a default size before the widget is available.
+        #then we set the default size on the widget and only use the widget size
+        self._default_size = p.size
+        self._init_glir()
+
+    def set_widget(self, widget):
+        self._widget = widget;
+        self._vispy_canvas.context.shared.parser._widget = widget;
```
Comment (rossant) [inline vispy/app/backends/_ipynb_webgl.py]: hey this is not JavaScript, no need to put a `;` :D

Comment (cgestes) [inline vispy/app/backends/_ipynb_webgl.py]: fixed

File: examples/ipynb/webgl_example_1.ipynb (hunk 1 above)
Comment (cgestes) [inline examples/ipynb/webgl_example_1.ipynb]: no idea how... wait... let me try something.

**Comment by rossant** [comment]
Seems to work!

* You'll need to fix flake/pep8 issues.
* Do we need to update vispy.min.js that is built into this repo?

File: examples/ipynb/webgl_example_1.ipynb (hunk 1 above)
Comment (cgestes) [inline examples/ipynb/webgl_example_1.ipynb]: done :)

**Comment by cgestes** [comment]
no need to update vispy.min.js

**Comment by rossant** [comment]
This is good to go as soon as it's flakified and travis is happy

**Comment by rossant** [comment]
@Eric89GXL we want to disable the test for ipynb_webgl backend (currently it fails because of IPython dependency). How to do that?

**Comment by Eric89GXL** [comment]
Change the "testable" variable for the backend

**Comment by rossant** [comment]
@cgestes can you try that?

**Comment by cgestes** [comment]
Already done, but I need to make the code importable when the import are not there.

Would prefer to define the testable state from outside the file I dont want to test...

**Comment by rossant** [comment]
works okay but I have a small problem with example 1: I cannot zoom with the wheel. No problem with example 2, weird... @cgestes can you reproduce this?

**Comment by cgestes** [comment]
none is working anymore.

**Comment by cgestes** [comment]
I think it's a vispy.js bug

**Comment by cgestes** [comment]
should be fixed now. could you try again please?

File: vispy/app/backends/ipython/__init__.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2014, 2015, Vispy Development Team.
+# Distributed under the (new) BSD License. See LICENSE.txt for more info.
+
+from ._widget import VispyWidget  # NOQA
+
+__all__ = ("VispyWidget")
```
Comment (rossant) [inline vispy/app/backends/ipython/__init__.py]: should this be a list instead?

File: vispy/app/backends/ipython/_widget.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+try:
+    from IPython.html.widgets import DOMWidget
+    from IPython.utils.traitlets import Unicode, Int, Bool
+except Exception as exp:
+    # Init dummy objects needed to import this module withour errors.
+    # These are all overwritten with imports from IPython (on success)
+    DOMWidget = object
+    Unicode = Int = Float = Bool = lambda *args, **kwargs: None
+
+    # raise ImportError("The WebGL backend requires IPython >= 2.0")
```
Comment (rossant) [inline vispy/app/backends/ipython/_widget.py]: remove this comment?

**Comment by rossant** [comment]
Works for me! LGTM except tiny comments.

File: vispy/app/backends/ipython/__init__.py (hunk 3 above)
Comment (cgestes) [inline vispy/app/backends/ipython/__init__.py]: good catch

**Comment by rossant** [comment]
thanks @cgestes 
awaiting for travis
this is ready to go, @Eric89GXL or @almarklein feel free to do a last check & merge

**Comment by Eric89GXL** [comment]
I don't know too much about this bit of code, I'll let @almarklein take a look if he has time. If he doesn't have time I'll muddle my way through it.

**Comment by Eric89GXL** [comment]
Ahh I just re-read your comment -- I didn't realize you had looked thoroughly already @rossant, I'll go ahead and merge then

**Comment by Eric89GXL** [comment]
(changes looked reasonable to me :) )

**Comment by rossant** [comment]
Thanks @Eric89GXL and @cgestes ! Now the WebGL backend should be a bit more robust. :)

**Pull request merged**